In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('horse_data.csv', header = None, na_values = '?')
df.reset_index()
df.columns = ['Surgery?','Age','Hospital_Number','rectal_temperature','pulse','respiratory_rate','temperature_of_extremities','peripheral_pulse','mucous_membranes','capillary_refill_time','pain','peristalsis','abdominal_distension','nasogastric_tube','nasogastric_reflux','nasogastric_reflux_PH','rectal_examination_feces','abdomen','packed_cell_volume','total_protein','abdominocentesis_appearance','abdomcentesis_total_protein','outcome','surgical_lesion?','site_of_lesion','type','subtype','cp_data']
df[df.pulse >148]

Surgery?  Age  Hospital_Number  rectal_temperature  pulse  \
3         1.0    9          5290409                39.1  164.0   
41        2.0    9          5288249                39.0  150.0   
55        1.0    9          5282839                38.6  160.0   
255       1.0    9          5294539                38.8  184.0   
275       1.0    9          5297159                38.8  150.0   

     respiratory_rate  temperature_of_extremities  peripheral_pulse  \
3                84.0                         4.0               1.0   
41               72.0                         NaN               NaN   
55               20.0                         3.0               NaN   
255              84.0                         1.0               NaN   
275              50.0                         1.0               3.0   

     mucous_membranes  capillary_refill_time  ...  packed_cell_volume  \
3                 6.0                    2.0  ...                48.0   
41                NaN                    NaN  ...                47.0   
55                5.0                    1.0  ...                38.0   
255               1.0                    1.0  ...                33.0   
275               6.0                    2.0  ...                50.0   

     total_protein  abdominocentesis_appearance  abdomcentesis_total_protein  \
3              7.2                          3.0                          5.3   
41             8.5                          NaN                          0.1   
55             NaN                          2.0                          NaN   
255            3.3                          NaN                          NaN   
275            6.2                          NaN                          NaN   

     outcome  surgical_lesion?  site_of_lesion  type  subtype  cp_data  
3        2.0                 1            2208     0        0        1  
41       1.0                 1            9400     0        0        1  
55       2.0                 1            7111     0        0        1  
255      2.0                 1            7111     0        0        2  
275      2.0                 1            4207     0        0        2  

[5 rows x 28 columns]

In [3]:
# Формируем DF из 8-ми столбцов
consize_df = df[['pulse','pain','rectal_temperature','temperature_of_extremities','abdominal_distension','mucous_membranes','peripheral_pulse','capillary_refill_time']]
consize_df.head()

pulse  pain  rectal_temperature  temperature_of_extremities  \
0   66.0   5.0                38.5                         3.0   
1   88.0   3.0                39.2                         NaN   
2   40.0   3.0                38.3                         1.0   
3  164.0   2.0                39.1                         4.0   
4  104.0   NaN                37.3                         NaN   

   abdominal_distension  mucous_membranes  peripheral_pulse  \
0                   4.0               NaN               3.0   
1                   2.0               4.0               NaN   
2                   1.0               3.0               1.0   
3                   4.0               6.0               1.0   
4                   NaN               6.0               NaN   

   capillary_refill_time  
0                    2.0  
1                    1.0  
2                    1.0  
3                    2.0  
4                    2.0

In [4]:
#Считаем стандартные характеристики через describe (ниже будут по отдельности)
consize_df.describe()

pulse       pain  rectal_temperature  temperature_of_extremities  \
count  276.000000  245.00000          240.000000                  244.000000   
mean    71.913043    2.95102           38.167917                    2.348361   
std     28.630557    1.30794            0.732289                    1.045054   
min     30.000000    1.00000           35.400000                    1.000000   
25%     48.000000    2.00000           37.800000                    1.000000   
50%     64.000000    3.00000           38.200000                    3.000000   
75%     88.000000    4.00000           38.500000                    3.000000   
max    184.000000    5.00000           40.800000                    4.000000   

       abdominal_distension  mucous_membranes  peripheral_pulse  \
count            244.000000        253.000000        231.000000   
mean               2.266393          2.853755          2.017316   
std                1.065131          1.620294          1.042428   
min                1.000000          1.000000          1.000000   
25%                1.000000          1.000000          1.000000   
50%                2.000000          3.000000          2.000000   
75%                3.000000          4.000000          3.000000   
max                4.000000          6.000000          4.000000   

       capillary_refill_time  
count             268.000000  
mean                1.305970  
std                 0.477629  
min                 1.000000  
25%                 1.000000  
50%                 1.000000  
75%                 2.000000  
max                 3.000000

In [5]:
#Посчитаем минимум, максимум и размах для количественных характеристик (для категориальных - не имеет смысла)
print(f'Максимум по столбцу Пульс:',consize_df.pulse.max())
print(f'Максимум по столбцу Ректальная температура:',consize_df.rectal_temperature.max())
print(f'Минимум по столбцу Пульс:',consize_df.pulse.min())
print(f'Минимум по столбцу Ректальная температура:',consize_df.rectal_temperature.min())
print(f'Размах по столбцу Пульс:',consize_df.pulse.max()-consize_df.pulse.min())
print('Размах по столбцу Ректальная температура: %.1f' % (consize_df.rectal_temperature.max()-consize_df.rectal_temperature.min()))

Максимум по столбцу Пульс: 184.0
Максимум по столбцу Ректальная температура: 40.8
Минимум по столбцу Пульс: 30.0
Минимум по столбцу Ректальная температура: 35.4
Размах по столбцу Пульс: 154.0
Размах по столбцу Ректальная температура: 5.4


In [6]:
#Посчитаем моду для категориальных величин
print(f'Мода по столбцу Боль:', consize_df.pain.mode()[0])
print(f'Мода по столбцу Температура конечностей:', consize_df.temperature_of_extremities.mode()[0])
print(f'Мода по столбцу Вздутие живота:', consize_df.abdominal_distension.mode()[0])
print(f'Мода по столбцу Слизистые оболочки:', consize_df.mucous_membranes.mode()[0])
print(f'Мода по столбцу Периферический пульс:', consize_df.peripheral_pulse.mode()[0])
print(f'Мода по столбцу Время наполнения капилляров:', consize_df.capillary_refill_time.mode()[0])

Мода по столбцу Боль: 3.0
Мода по столбцу Температура конечностей: 3.0
Мода по столбцу Вздутие живота: 1.0
Мода по столбцу Слизистые оболочки: 1.0
Мода по столбцу Периферический пульс: 1.0
Мода по столбцу Время наполнения капилляров: 1.0


In [7]:
#Посчитаем меры центральной тенденции для количественных характеристик  (для категориальных - медиана и среднее не имеет смысла)
print(f'Мода по столбцу Пульс: %.1f' %consize_df.pulse.mode())
print(f'Мода по столбцу Ректальная температура: %.1f' %consize_df.rectal_temperature.mode())
print(f'Медиана по столбцу Пульс: %.1f' %consize_df.pulse.median())
print(f'Медиана по столбцу Ректальная температура: %.1f' %consize_df.rectal_temperature.median())
print(f'Среднее арифметическое по столбцу Пульс: %.1f' % consize_df.pulse.mean())
print(f'Среднее арифметическое по столбцу Ректальная температура: %.1f' % consize_df.rectal_temperature.mean())

Мода по столбцу Пульс: 48.0
Мода по столбцу Ректальная температура: 38.0
Медиана по столбцу Пульс: 64.0
Медиана по столбцу Ректальная температура: 38.2
Среднее арифметическое по столбцу Пульс: 71.9
Среднее арифметическое по столбцу Ректальная температура: 38.2


In [8]:
#Посчитаем 0,25 и 0,75 квартили и межквартильный размах для количественных характеристик (для категориальных - не имеет смысла)
print(f'0,25 квартиль по столбцу Пульс: %.1f' %consize_df.pulse.quantile([0.25]))
print(f'0,75 квартиль по столбцу Пульс: %.1f' %consize_df.pulse.quantile([0.75]))
print(f'0,25 квартиль по столбцу Ректальная температура: %.1f' %consize_df.rectal_temperature.quantile([0.25]))
print(f'0,75 квартиль по столбцу Ректальная температура: %.1f' %consize_df.rectal_temperature.quantile([0.75]))
print(f'Межквартильный размах по столбцу Пульс: %.1f' %(consize_df.pulse.quantile(0.75)-consize_df.pulse.quantile(0.25)))
print(f'Межквартильный размах по столбцу Ректальная температура: %.1f' %(consize_df.rectal_temperature.quantile(0.75)-consize_df.rectal_temperature.quantile(0.25)))

0,25 квартиль по столбцу Пульс: 48.0
0,75 квартиль по столбцу Пульс: 88.0
0,25 квартиль по столбцу Ректальная температура: 37.8
0,75 квартиль по столбцу Ректальная температура: 38.5
Межквартильный размах по столбцу Пульс: 40.0
Межквартильный размах по столбцу Ректальная температура: 0.7


In [9]:
# Посчитаем СКО и дисперсию (для категориальных - не имеет смысла)
print(f'СКО по столбцу Пульс: %.2f' %consize_df.pulse.std())
print(f'СКО по столбцу Ректальная температура: %.2f' %consize_df.rectal_temperature.std())
print(f'Дисперсия по столбцу Пульс: %.2f' %consize_df.pulse.var())
print(f'Дисперсия по столбцу Ректальная температура: %.2f' %consize_df.rectal_temperature.var())
print(f'Наличие выбросов слева по по столбцу Пульс: %.2f' %(consize_df.pulse.quantile(0.25) - 1.5*(consize_df.pulse.quantile(0.75)-consize_df.pulse.quantile(0.25))))
print(f'Наличие выбросов справа по по столбцу Пульс: %.2f' %(consize_df.pulse.quantile(0.75) + 1.5*(consize_df.pulse.quantile(0.75)-consize_df.pulse.quantile(0.25))))
print(f'Наличие выбросов слева по по столбцу Ректальная температура: %.2f' %(consize_df.rectal_temperature.quantile(0.25) - 1.5*(consize_df.rectal_temperature.quantile(0.75)-consize_df.rectal_temperature.quantile(0.25))))
print(f'Наличие выбросов справа по по столбцу Ректальная температура: %.2f' %(consize_df.rectal_temperature.quantile(0.75) + 1.5*(consize_df.rectal_temperature.quantile(0.75)-consize_df.rectal_temperature.quantile(0.25))))

СКО по столбцу Пульс: 28.63
СКО по столбцу Ректальная температура: 0.73
Дисперсия по столбцу Пульс: 819.71
Дисперсия по столбцу Ректальная температура: 0.54
Наличие выбросов слева по по столбцу Пульс: -12.00
Наличие выбросов справа по по столбцу Пульс: 148.00
Наличие выбросов слева по по столбцу Ректальная температура: 36.75
Наличие выбросов справа по по столбцу Ректальная температура: 39.55


Задание 1
Значение всех показателей центральной тенденции для ректальной температуры несколько выше нормального значения температуры для лошадей (указана в спецификации 37,8), однако разница между этими значениями (мода, медиана, среднее ариф.) незначительна - не более 0,2. По данному показателю судить о состоянии здоровья лошадей в выборке в целом нельзя - слишком небольшое отклонение от нормы. Но при этом говорить о том, что больных лошадей нет - мы не можем, все-таки все значения показателей центральной тенденции выше нормы, а значит больные лошади есть. Собственно, это видно из сравнений квартилей 0,25 квартиль = 37,8 - т.е. только 25% лошадей из выборки имели нормальную температуру.Однако у следующей половины лошадей (0,75 - 0,25) температура была повышена не более чем на 1 градус (0,8). При проверке на наличие выбросов (через 1.5 IOR) справа и слева наблюдаются выбросы (min < 36,75 и max > 39.55) - поэтому необходима проверка на корректность данных столбца.

Для значений пульса значение моды (48) не очень показательно, так как размах для этого показателя намного более выше, чем само значение моды. Медиана равна 64, а среднее 71,9 - т.е. распределение смещено вправо. Эта же закономерность видна при сравнении квартилей (Q2-Q1)= 16, а (Q3-Q2)=24. Однако, все показатели центральной тенденции показали значительное отклонение от нормы (указана в спецификации 30-40), что говорит о том, что в большинстве случаев в выборке были лошади, имеющие проблемы со здоровьем. Также хотелось бы обратить внимание на то, что слева у данного показателя выбросов нет - минимум (30) - норма для лошади (возможно, для молодой возможно - но данных о норме пульса для молодняка - в описании нет). Максимальное значение пульса 184 - что значительно выше даже среднего значения. При проверке на наличие выбросов (через 1.5 IOR) выбросы наблюдаются только справа (max > 148).Поэтому конечно, нельзя не исключить наличие выбросов справа у показателя пульса. 

Количественных характеристик в сокращенном датафрейме больше нет. Для категориальных характеристик имеет смысл смотреть только моду.
По показателю Боль - чаще всего наблюдалась периодическая легкая боль
По показателю Температура конечностей - чаще всего наблюдались прохладные ноги, возможно у этих лошадей возможен шок.
По показателю Вздутие живота - чаще всего вздутие живота не наблюдалось
По показателю Слизистые оболочки - чаще всего у лошадей наблюдалась норма - они нормального розового цвета
По показателю Периферический пульс - чаще всего вздутие живота наблюдалась норма
По показателю Время наполнения капилляров - чаще всего нарушение кровообращения (указывает на шок) - отклонение от нормы

In [10]:
# Для анализа выбросов по высокому пульсу
consize_df[consize_df.pulse >148]

pulse  pain  rectal_temperature  temperature_of_extremities  \
3    164.0   2.0                39.1                         4.0   
41   150.0   NaN                39.0                         NaN   
55   160.0   3.0                38.6                         3.0   
255  184.0   4.0                38.8                         1.0   
275  150.0   5.0                38.8                         1.0   

     abdominal_distension  mucous_membranes  peripheral_pulse  \
3                     4.0               6.0               1.0   
41                    NaN               NaN               NaN   
55                    4.0               5.0               NaN   
255                   3.0               1.0               NaN   
275                   2.0               6.0               3.0   

     capillary_refill_time  
3                      2.0  
41                     NaN  
55                     1.0  
255                    1.0  
275                    2.0

In [11]:
# Для анализа выбросов по низкой температуре
consize_df[consize_df.rectal_temperature < 36.75]

pulse  pain  rectal_temperature  temperature_of_extremities  \
44   140.0   4.0                35.4                         3.0   
80    98.0   4.0                36.4                         3.0   
118   78.0   5.0                36.5                         1.0   
141   42.0   NaN                36.0                         NaN   
238   88.0   3.0                36.1                         3.0   
251   42.0   1.0                36.6                         3.0   
298  100.0   3.0                36.5                         3.0   

     abdominal_distension  mucous_membranes  peripheral_pulse  \
44                    NaN               4.0               3.0   
80                    2.0               4.0               3.0   
118                   1.0               1.0               NaN   
141                   NaN               5.0               NaN   
238                   2.0               3.0               3.0   
251                   1.0               2.0               3.0   
298                   3.0               3.0               3.0   

     capillary_refill_time  
44                     2.0  
80                     1.0  
118                    1.0  
141                    1.0  
238                    1.0  
251                    1.0  
298                    1.0

In [12]:
# Для анализа выбросов по высокой температуре
consize_df[consize_df.rectal_temperature > 39.55]

pulse  pain  rectal_temperature  temperature_of_extremities  \
20    72.0   5.0                39.9                         1.0   
54   114.0   2.0                40.3                         3.0   
75   100.0   2.0                39.7                         3.0   
91   114.0   2.0                40.3                         3.0   
99   108.0   2.0                39.6                         3.0   
259   72.0   2.0                40.8                         3.0   
281   78.0   2.0                40.0                         3.0   

     abdominal_distension  mucous_membranes  peripheral_pulse  \
20                    4.0               5.0               1.0   
54                    3.0               1.0               3.0   
75                    NaN               5.0               3.0   
91                    3.0               1.0               3.0   
99                    3.0               6.0               3.0   
259                   1.0               1.0               3.0   
281                   1.0               5.0               3.0   

     capillary_refill_time  
20                     2.0  
54                     2.0  
75                     2.0  
91                     2.0  
99                     2.0  
259                    1.0  
281                    1.0

Задание 2 Корректность выбросов можно проверить на основе корреляции с другими данными: 
    Для Пульса: 
        Подозрение на выбросы только справа. Согласно данным из спецификации, значения пульса могут быть увеличены если лошадь:
            - испытывает боль (у index = 275 - самая сильная боль - это не выброс, оставляем в выборке
            Помимо этого у лошади с index= 255 - также значительное вздутие живота, что подтверждает оценку высокой боли - оставляем
            Слизистые оболочки у лошади с index= 275 - темно-синюшные, также у лошади понижен периферический пульс, что говорит о пониженном пульсе 
            - если есть шок кровообращения - на наличие шока может указывать холодные конечности - это лошади с index = 3;55. 
            Также у этих лошадей наблюдается сильное вздутие живота, для вздутия характерна сильная боль.
            Поэтому учитывая субъективность оценки боли, скорее показатель боли недооценен, а показатель пульса корректен - оставляем в выборке.
            - при повышенной температуре может наблюдаться шок кровообращения - также при шоке кровообращения может быть повышенный пульс, что наблюдается
            для лошади с index= 41 - оставляем
    Для Ректальной температуры: 
        Подозрение на выбросы и справа, и слева:
             Согласно данным из спецификации, значения ректальной температуры могут быть снижены если лошадь: 
            - испытывает шок, на шок также указывает:
                - холодные конечности - нет у лошадей с выбросами
                - бледно-розовые слизистые оболочки 238;298
                - повышенный пульс есть у index = 44;80;118;238;298
                - высокая боль есть у index = 44;80;118
                - пониженное время заполнения капилляров есть у index = 80;118;141;238;251;298
            для сепсиса 
                *лошадей с index = 44;80;118;238;251;298 ввиду пересечения с несколькими показателями.
            для сепсиса (подозревается для лошади 141 - ярко-красные слизистые и пониженное время заполнения капилляров) 
                возможна как высокая, так и низкая температура, но при этом должен быть повышенный пульс, 
                а его нет (он почти в указанной норме). также ввиду отсутствия других показателей нельзя принять 
                однозначное решение, какой из показателей неверно установлен (пульс или ректальная температура)
                    - приняла решение удалить лошадь с index = 141 из выборки           
            Согласно данным из спецификации, значения ректальной температуры могут быть повышены если у лошади: 
            - есть инфекция
            - для лошадей с index = 54;75;91;99;259;281 - оставляем данные, это не выбросы. У этих лошадей
                температура конечностей также повышена
            - для лошади с index = 20 температура конечностей не повышена, и на кажущийся в первую очередь выброс,
                я приняла решение оставить это значение из-за: цвета слизистых оболочек (указывают на сепсис,(это инфекция
                - при сепсисе также характерна высокая температура)
                                                                                       

In [13]:
# убираем из выборки index = 141
cleaned_df = consize_df.drop(consize_df.index[141])

In [14]:
# Пересчитаем стандартные характеристики:
cleaned_df.describe()

pulse       pain  rectal_temperature  temperature_of_extremities  \
count  275.000000  245.00000          239.000000                  244.000000   
mean    72.021818    2.95102           38.176987                    2.348361   
std     28.625563    1.30794            0.720187                    1.045054   
min     30.000000    1.00000           35.400000                    1.000000   
25%     48.000000    2.00000           37.800000                    1.000000   
50%     64.000000    3.00000           38.200000                    3.000000   
75%     88.000000    4.00000           38.500000                    3.000000   
max    184.000000    5.00000           40.800000                    4.000000   

       abdominal_distension  mucous_membranes  peripheral_pulse  \
count            244.000000        252.000000        231.000000   
mean               2.266393          2.845238          2.017316   
std                1.065131          1.617834          1.042428   
min                1.000000          1.000000          1.000000   
25%                1.000000          1.000000          1.000000   
50%                2.000000          3.000000          2.000000   
75%                3.000000          4.000000          3.000000   
max                4.000000          6.000000          4.000000   

       capillary_refill_time  
count             267.000000  
mean                1.307116  
std                 0.478156  
min                 1.000000  
25%                 1.000000  
50%                 1.000000  
75%                 2.000000  
max                 3.000000

In [15]:
print(f'Максимум по столбцу Ректальная температура:',cleaned_df.rectal_temperature.max())
print(f'Минимум по столбцу Ректальная температура:',cleaned_df.rectal_temperature.min())
print('Размах по столбцу Ректальная температура: %.1f' % (cleaned_df.rectal_temperature.max()-cleaned_df.rectal_temperature.min()))
print(f'Мода по столбцу Ректальная температура: %.1f' %cleaned_df.rectal_temperature.mode())
print(f'Медиана по столбцу Ректальная температура: %.1f' %cleaned_df.rectal_temperature.median())
print(f'Среднее арифметическое по столбцу Ректальная температура: %.1f' % cleaned_df.rectal_temperature.mean())
print(f'0,25 квартиль по столбцу Ректальная температура: %.1f' %cleaned_df.rectal_temperature.quantile([0.25]))
print(f'0,75 квартиль по столбцу Ректальная температура: %.1f' %cleaned_df.rectal_temperature.quantile([0.75]))
print(f'Межквартильный размах по столбцу Ректальная температура: %.1f' %(cleaned_df.rectal_temperature.quantile(0.75)-cleaned_df.rectal_temperature.quantile(0.25)))
print(f'СКО по столбцу Ректальная температура: %.2f' %cleaned_df.rectal_temperature.std())
print(f'Дисперсия по столбцу Ректальная температура: %.2f' %cleaned_df.rectal_temperature.var())

Максимум по столбцу Ректальная температура: 40.8
Минимум по столбцу Ректальная температура: 35.4
Размах по столбцу Ректальная температура: 5.4
Мода по столбцу Ректальная температура: 38.0
Медиана по столбцу Ректальная температура: 38.2
Среднее арифметическое по столбцу Ректальная температура: 38.2
0,25 квартиль по столбцу Ректальная температура: 37.8
0,75 квартиль по столбцу Ректальная температура: 38.5
Межквартильный размах по столбцу Ректальная температура: 0.7
СКО по столбцу Ректальная температура: 0.72
Дисперсия по столбцу Ректальная температура: 0.52


In [16]:
# Задание 3
cleaned_df.info()
for col in cleaned_df.columns:
    pct_missing = cleaned_df[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 299 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   pulse                       275 non-null    float64
 1   pain                        245 non-null    float64
 2   rectal_temperature          239 non-null    float64
 3   temperature_of_extremities  244 non-null    float64
 4   abdominal_distension        244 non-null    float64
 5   mucous_membranes            252 non-null    float64
 6   peripheral_pulse            231 non-null    float64
 7   capillary_refill_time       267 non-null    float64
dtypes: float64(8)
memory usage: 21.0 KB
pulse - 8.0%
pain - 18.1%
rectal_temperature - 20.1%
temperature_of_extremities - 18.4%
abdominal_distension - 18.4%
mucous_membranes - 15.7%
peripheral_pulse - 22.7%
capillary_refill_time - 10.7%


Задание 3 - предлагаемая методика работы с пропусками
Столбцы с пропусками не стоит удалять из DF, так как присутствие данных по всех столбцах достаточно (пропусков не более 25%).

Я подозреваю, что наличие пропусков в первую очередь связано с двумя причинами:
    1) лошадь поступила в тяжелом состоянии - все обследование сделать не получилось или лошадь умерла и обследование закончить не успели
    2) при обследовании получили информацию, что позволило поставить диагноз и остальные исследования не имели смысла.
Я думаю, что эти пропуски относятся к MAR, поэтому замена пропусков не должна привести к существенному искажению результатов.

Попробую заменить пропуски и уже потом при наличии строк с пустыми значениями удалю их (при недостаточности данных в этой строке для относительно корректной замены значений)

In [17]:
#Заполним пропуски по ректальной температуре медианой (только для случаев с повышенной температурой конечностей) - "Имитация жара"
high_temp_ex = cleaned_df['temperature_of_extremities'] == 2.0
median_high_temp_ex = cleaned_df[high_temp_ex]['rectal_temperature'].median()
cleaned_df.loc[high_temp_ex & cleaned_df.rectal_temperature.isna(),'rectal_temperature'] = median_high_temp_ex

In [18]:
# Заполним пропуски по ректальной температуре медианой по категориям температуры конечностей и времени наполнения капилляров:
cleaned_df.rectal_temperature.fillna(cleaned_df.groupby(['temperature_of_extremities', 'capillary_refill_time']).rectal_temperature.transform('median'), inplace=True)

In [19]:
# Заполним пропуски по ректальной температуре медианой по категории времени наполнения капилляров или температуры конечностей:
cleaned_df.rectal_temperature.fillna(cleaned_df.groupby(['capillary_refill_time']).rectal_temperature.transform('median'), inplace=True)
cleaned_df.rectal_temperature.fillna(cleaned_df.groupby(['temperature_of_extremities']).rectal_temperature.transform('median'), inplace=True)

In [20]:
# Заполним пропуски по пульсу медианой для темно-синюшных слизистых, пониженного периферического пульса, холодных конечностей "Имитация шока":
low_temp_ex = cleaned_df['temperature_of_extremities'] == (4.0 or 3.0)
low_pp = cleaned_df['peripheral_pulse'] == (3.0 or 4.0)
dark_cyan = cleaned_df['mucous_membranes'] == 6.0

cleaned_df.loc[low_temp_ex & low_pp & dark_cyan & cleaned_df.pulse.isna(),'pulse'] = cleaned_df[low_temp_ex & low_pp & dark_cyan]['pulse'].median()
cleaned_df.loc[low_temp_ex & low_pp & cleaned_df.pulse.isna(),'pulse'] = cleaned_df[low_temp_ex & low_pp]['pulse'].median()
cleaned_df.loc[low_pp & dark_cyan & cleaned_df.pulse.isna(),'pulse'] = cleaned_df[low_pp & dark_cyan]['pulse'].median()
cleaned_df.loc[low_temp_ex & dark_cyan & cleaned_df.pulse.isna(),'pulse'] = cleaned_df[low_temp_ex & dark_cyan]['pulse'].median()

In [21]:
#Заполним пропуски по пульсу (только для случаев с болью) - "Имитация боли"
high_pain = cleaned_df['pain'] == (4.0 or 5.0)
high_median = cleaned_df[(cleaned_df['pain'] == (4.0 or 5.0))]['pulse'].median()
cleaned_df.loc[high_pain & cleaned_df.pulse.isna(),'pulse'] = high_median

In [22]:
# Заполним пропуски по пульсу медианой по категориям (температура конечностей):
cleaned_df.pulse.fillna(cleaned_df.groupby(['temperature_of_extremities']).pulse.transform('median'), inplace=True)

In [23]:
high_temp = cleaned_df['rectal_temperature'] > 37.8
low_temp = cleaned_df['rectal_temperature'] <= 37.8
mode_high_temp = cleaned_df[high_temp]['temperature_of_extremities'].mode()[0]
mode_low_temp = cleaned_df[low_temp]['temperature_of_extremities'].mode()[0]

#Заполним пропуски по температуре конечностей модой (только для случаев с повышенной ректальной температурой) - "Имитация жара"
cleaned_df.loc[high_temp & cleaned_df.temperature_of_extremities.isna(),'temperature_of_extremities'] = mode_high_temp

#Заполним пропуски по температуре конечностей модой (только для случаев с пониженной ректальной температурой) - "Имитация шока"
cleaned_df.loc[low_temp & cleaned_df.temperature_of_extremities.isna(),'temperature_of_extremities'] = mode_low_temp

In [24]:
mode_cap_refill_1 = cleaned_df[high_temp]['capillary_refill_time'].mode()[0]
mode_cap_refill_2 = cleaned_df[low_temp]['capillary_refill_time'].mode()[0]
# Заполним пропуски по времени заполнения капилляров модой (только для случаев с повышенной ректальной температурой)
cleaned_df.loc[high_temp & cleaned_df.capillary_refill_time.isna(),'capillary_refill_time'] = mode_cap_refill_1
# Заполним пропуски по по времени заполнения капилляров модой (только для случаев с пониженной ректальной температурой)
cleaned_df.loc[low_temp & cleaned_df.capillary_refill_time.isna(),'capillary_refill_time'] = mode_cap_refill_2

In [25]:
# Заполним пропуски по периф. пульсу (корреляция с временем заполнения капилляров, слиз. оболочками):
# Корреляцию с пульсом не стала смотреть ввиду высокой дисперсии пульса.
cleaned_df['peripheral_pulse'] = cleaned_df.groupby(['capillary_refill_time','mucous_membranes'])['peripheral_pulse'].apply(lambda x: x.fillna(x.mode()[0]))
cleaned_df['peripheral_pulse'] = cleaned_df.groupby(['capillary_refill_time'])['peripheral_pulse'].apply(lambda x: x.fillna(x.mode()[0]))

In [26]:
# Заполним пропуски по слизистым оболочкам (корреляция с временем заполнения капилляров, периф пульсом):
# Корреляцию с пульсом не стала смотреть ввиду высокой дисперсии пульса.
cleaned_df['mucous_membranes'] = cleaned_df.groupby(['capillary_refill_time','peripheral_pulse'])['mucous_membranes'].apply(lambda x: x.fillna(x.mode()[0]))
cleaned_df['mucous_membranes'] = cleaned_df.groupby(['capillary_refill_time'])['mucous_membranes'].apply(lambda x: x.fillna(x.mode()[0]))

In [27]:
#Заполним пропуски по вздутию (только для случаев с высокой болью) - "Имитация боли"
high_pain4 = cleaned_df['pain'] == 4.0
high_pain5 = cleaned_df['pain'] == 5.0
high_pain_mode4 = cleaned_df[high_pain4]['abdominal_distension'].mode()[0]
high_pain_mode5 = cleaned_df[high_pain5]['abdominal_distension'].mode()[0]
cleaned_df.loc[high_pain4 & cleaned_df.abdominal_distension.isna(),'abdominal_distension'] = high_pain_mode4
cleaned_df.loc[high_pain5 & cleaned_df.abdominal_distension.isna(),'abdominal_distension'] = high_pain_mode5

In [28]:
mode_high_temp_pain = cleaned_df[high_temp]['pain'].mode()[0]
mode_low_temp_pain = cleaned_df[low_temp]['pain'].mode()[0]

#Заполним пропуски по температуре конечностей модой (только для случаев с повышенной ректальной температурой) - "Имитация жара"
cleaned_df.loc[high_temp & cleaned_df.pain.isna(),'pain'] = mode_high_temp_pain

#Заполним пропуски по температуре конечностей модой (только для случаев с пониженной ректальной температурой) - "Имитация шока"
cleaned_df.loc[low_temp & cleaned_df.pain.isna(),'pain'] = mode_low_temp_pain

In [29]:
#Заполним пропуски по вздутию модой для категории боли, периф пульса
cleaned_df['abdominal_distension'] = cleaned_df.groupby(['pain'])['abdominal_distension'].apply(lambda x: x.fillna(x.mode()[0]))

In [31]:
# Посмотрим, сколько пропусков осталось после манипуляций
for col in cleaned_df.columns:
    pct_missing = cleaned_df[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

pulse - 2.7%
pain - 3.0%
rectal_temperature - 3.3%
temperature_of_extremities - 3.3%
abdominal_distension - 3.0%
mucous_membranes - 3.3%
peripheral_pulse - 3.3%
capillary_refill_time - 3.3%


In [32]:
# Удалим оставшиеся пропуски в датафрейме
result = cleaned_df.dropna(how='any')
result.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 289 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   pulse                       289 non-null    float64
 1   pain                        289 non-null    float64
 2   rectal_temperature          289 non-null    float64
 3   temperature_of_extremities  289 non-null    float64
 4   abdominal_distension        289 non-null    float64
 5   mucous_membranes            289 non-null    float64
 6   peripheral_pulse            289 non-null    float64
 7   capillary_refill_time       289 non-null    float64
dtypes: float64(8)
memory usage: 20.3 KB
